File: packages/core/voxnote/pipeline/exporter.py
"""Export meeting insights as Obsidian-compatible Markdown."""

from __future__ import annotations

from datetime import datetime
from pathlib import Path

from rich.console import Console

from voxnote.config import settings
from voxnote.pipeline.models import TranscriptResult

console = Console()


def export_obsidian(
    insights: dict,
    transcript: str | TranscriptResult,
    audio_name: str,
    output_dir: Path | None = None,
) -> Path:
    """Generate an Obsidian Markdown note from meeting insights.

    Args:
        insights: Structured dict returned by :func:`extract_insights`.
        transcript: Full transcription text or TranscriptResult with speaker info.
        audio_name: Original audio filename (used in the note title).
        output_dir: Override for settings.output_dir.

    Returns:
        Path to the written Markdown file.
    """
    # Normalize transcript to string
    if isinstance(transcript, TranscriptResult):
        transcript_text = transcript.to_speaker_text()
    else:
        transcript_text = transcript

    output_dir = output_dir or settings.output_dir
    output_dir.mkdir(parents=True, exist_ok=True)

    slug = Path(audio_name).stem

    import re

    # Match format YYYY-MM-DD_HH-MM-SS
    match = re.match(r"^(\d{4}-\d{2}-\d{2})_(\d{2})-(\d{2})-(\d{2})_(.*)$", slug)
    if match:
        date_str = match.group(1)
        time_str = f"{match.group(2)}:{match.group(3)}"
        title_slug = match.group(5)
        note_filename = f"{slug}.md"
    else:
        now = datetime.now()
        date_str = now.strftime("%Y-%m-%d")
        time_str = now.strftime("%H:%M")
        title_slug = slug
        note_filename = f"{date_str}_{slug}.md"

    def _as_list(value: object) -> list:
        # LLM output sometimes returns a bare string where a list is expected. Iterating
        # that string would emit one bullet per character, so coerce to a list first.
        if value is None or value == "":
            return []
        return value if isinstance(value, list) else [value]

    def _bullets(values: object, empty: str) -> str:
        items = [str(v).strip() for v in _as_list(values) if str(v).strip()]
        return "\n".join(f"- {v}" for v in items) if items else f"- {empty}"

    tasks = ""
    for item in _as_list(insights.get("action_items")):
        if not isinstance(item, dict) or not item.get("task"):
            continue
        resp = item.get("owner", "TBD")
        dl = item.get("deadline", "TBD")
        tasks += f"- [ ] {item['task']} @{resp} (deadline: {dl})\n"

    participants_rows = []
    for p in _as_list(insights.get("participants")):
        if isinstance(p, dict) and p.get("speaker"):
            contribution = p.get("contribution", "")
            participants_rows.append(
                f"- **{p['speaker']}**" + (f" — {contribution}" if contribution else "")
            )
        elif isinstance(p, str) and p.strip():
            participants_rows.append(f"- **{p.strip()}**")

    highlights_rows = []
    for c in _as_list(insights.get("highlights")):
        if isinstance(c, dict) and c.get("quote"):
            speaker = c.get("speaker", "")
            prefix = f"**{speaker}:** " if speaker else ""
            highlights_rows.append(f"> {prefix}{c['quote']}")
        elif isinstance(c, str) and c.strip():
            highlights_rows.append(f"> {c.strip()}")

    decisions = _bullets(insights.get("decisions", []), "None")
    insight_lines = _bullets(insights.get("insights", []), "None")
    questions = _bullets(insights.get("open_questions", []), "None")
    next_steps = _bullets(insights.get("next_steps", []), "None")
    tasks_md = tasks.rstrip() if tasks else "- [ ] No tasks identified"

    sections = [f"## 📝 Summary\n\n{insights.get('summary', 'N/A')}"]
    if participants_rows:
        sections.append("## 👥 Participants\n\n" + "\n".join(participants_rows))
    if insights.get("key_points"):
        sections.append("## 📌 Key points\n\n" + _bullets(insights["key_points"], ""))
    sections.append("## ✅ Decisions\n\n" + decisions)
    sections.append("## 🎯 Action items\n\n" + tasks_md)
    sections.append("## 💡 Insights\n\n" + insight_lines)
    if highlights_rows:
        sections.append("## 💬 Highlights\n\n" + "\n".join(highlights_rows))
    sections.append("## ❓ Open questions\n\n" + questions)
    sections.append("## 🔜 Next steps\n\n" + next_steps)
    body = "\n\n---\n\n".join(sections)

    note = f"""\
---
tags: [meeting, {date_str}]
date: {date_str}
time: {time_str}
audio: "[[audio/{Path(audio_name).name}]]"
---

# 📋 Meeting — {title_slug}

> 🗓️ **Date:** {date_str} · ⏰ **Time:** {time_str} · 🎧 `{Path(audio_name).name}`

---

{body}

---

<details>
<summary>📄 Full transcript</summary>

{transcript_text}

</details>
"""

    note_path = output_dir / note_filename
    note_path.write_text(note, encoding="utf-8")
    try:
        note_path.chmod(0o600)  # notes can contain sensitive meeting content
    except OSError:
        pass
    console.print(f"[green]Note saved[/] → {note_path}")
    return note_path

File: packages/web/src/lib/exporter.ts
/**
 * Browser-side note generation — a faithful TypeScript port of the Python
 * `voxnote.pipeline.exporter.export_obsidian` so notes produced in the web (no backend)
 * are byte-identical in structure to the ones the local CLI writes, and stay
 * Obsidian-compatible.
 *
 * Differences from the Python version:
 *  - No filesystem: returns `{ filename, content }` instead of writing a file.
 *  - `audio_filename` is the original upload/recording name (no server-side path).
 */

import type { InsightsResult, ExportResult } from "@/types";

function asList(value: unknown): unknown[] {
  if (value === null || value === undefined || value === "") return [];
  return Array.isArray(value) ? value : [value];
}

function bullets(values: unknown, empty: string): string {
  const items = asList(values)
    .map((v) => String(v).trim())
    .filter((v) => v.length > 0);
  return items.length ? items.map((v) => `- ${v}`).join("\n") : `- ${empty}`;
}

export interface ExportNoteOptions {
  audioFilename?: string;
}

export function exportNote(
  transcript: string,
  insights: InsightsResult,
  audioFilename: string = "recording.wav",
): ExportResult {
  const slug = audioFilename.replace(/\.[^.]+$/, "");

  // Match the Python slug format YYYY-MM-DD_HH-MM-SS_title
  const match = slug.match(/^(\d{4}-\d{2}-\d{2})_(\d{2})-(\d{2})-(\d{2})_(.*)$/);
  const now = new Date();
  let dateStr: string;
  let timeStr: string;
  let titleSlug: string;
  let noteFilename: string;
  if (match) {
    dateStr = match[1];
    timeStr = `${match[2]}:${match[3]}`;
    titleSlug = match[5];
    noteFilename = `${slug}.md`;
  } else {
    const pad = (n: number) => String(n).padStart(2, "0");
    dateStr = `${now.getFullYear()}-${pad(now.getMonth() + 1)}-${pad(now.getDate())}`;
    timeStr = `${pad(now.getHours())}:${pad(now.getMinutes())}`;
    titleSlug = slug;
    noteFilename = `${dateStr}_${slug}.md`;
  }

  // Action items → task list with owner/deadline
  let tasks = "";
  for (const item of asList(insights.action_items)) {
    if (typeof item !== "object" || item === null) continue;
    const it = item as { task?: string; owner?: string; deadline?: string };
    if (!it.task) continue;
    const owner = it.owner || "TBD";
    const deadline = it.deadline || "TBD";
    tasks += `- [ ] ${it.task} @${owner} (deadline: ${deadline})\n`;
  }
  const tasksMd = tasks.trim() || "- [ ] No tasks identified";

  const participantsRows: string[] = [];
  for (const p of asList(insights.participants)) {
    if (typeof p === "object" && p !== null) {
      const pp = p as { speaker?: string; contribution?: string };
      if (pp.speaker) {
        participantsRows.push(
          `**${pp.speaker}**` + (pp.contribution ? ` — ${pp.contribution}` : ""),
        );
      }
    } else if (typeof p === "string" && p.trim()) {
      participantsRows.push(`**${p.trim()}**`);
    }
  }

  const highlightsRows: string[] = [];
  for (const c of asList(insights.highlights)) {
    if (typeof c === "object" && c !== null) {
      const cc = c as { speaker?: string; quote?: string };
      if (cc.quote) {
        const prefix = cc.speaker ? `**${cc.speaker}:** ` : "";
        highlightsRows.push(`> ${prefix}${cc.quote}`);
      }
    } else if (typeof c === "string" && c.trim()) {
      highlightsRows.push(`> ${c.trim()}`);
    }
  }

  const decisions = bullets(insights.decisions, "None");
  const insightLines = bullets(insights.insights, "None");
  const questions = bullets(insights.open_questions, "None");
  const nextSteps = bullets(insights.next_steps, "None");

  const sections: string[] = [`## 📝 Summary\n\n${insights.summary || "N/A"}`];
  if (participantsRows.length) sections.push(`## 👥 Participants\n\n${participantsRows.join("\n")}`);
  if (insights.key_points && insights.key_points.length) {
    sections.push(`## 📌 Key points\n\n${bullets(insights.key_points, "")}`);
  }
  sections.push(`## ✅ Decisions\n\n${decisions}`);
  sections.push(`## 🎯 Action items\n\n${tasksMd}`);
  sections.push(`## 💡 Insights\n\n${insightLines}`);
  if (highlightsRows.length) sections.push(`## 💬 Highlights\n\n${highlightsRows.join("\n")}`);
  sections.push(`## ❓ Open questions\n\n${questions}`);
  sections.push(`## 🔜 Next steps\n\n${nextSteps}`);
  const body = sections.join("\n\n---\n\n");

  const audioName = audioFilename.split("/").pop() || audioFilename;

  const content = `---
tags: [meeting, ${dateStr}]
date: ${dateStr}
time: ${timeStr}
audio: "[[audio/${audioName}]]"
---

# 📋 Meeting — ${titleSlug}

> 🗓️ **Date:** ${dateStr} · ⏰ **Time:** ${timeStr} · 🎧 \`${audioName}\`

---

${body}

---

<details>
<summary>📄 Full transcript</summary>

${transcript}

</details>
`;

  // path is irrelevant in the browser; kept for type compatibility with ExportResult.
  return { filename: noteFilename, content, path: noteFilename };
}
